Helm and updates

dereknorrbom opened the pull request:
- Add Helm chart for Kubernetes deployment
- Create setup.sh script to install dependencies (kubectl, Helm, Docker)
- Implement Go integration tests for Kubernetes interactions
- Update Dockerfile to use distroless base image
- Refactor main.go to use Kubernetes lister for efficient caching
- Enhance Makefile with new targets for Helm deployment and testing, refactor for DRY

dereknorrbom committed:
Helm and updates

- Add Helm chart for Kubernetes deployment
- Create setup.sh script to install dependencies (kubectl, Helm, Docker)
- Implement Go integration tests for Kubernetes interactions
- Update Dockerfile to use distroless base image
- Refactor main.go to use Kubernetes lister for efficient caching
- Enhance Makefile with new targets for Helm deployment and testing, refactor for DRY

taraspos commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -25,26 +28,22 @@ import (
+}
 
 // main initializes and runs the server, setting up TLS configuration, handling graceful shutdown, and defining HTTP routes.
 func main() {
```
Your `main.go` file now quite bloated with all kinds of logic.
Have you considered splitting it to more packages, or files for a bit better logical organization?

dereknorrbom replied:
Yes. I listed it as a todo in a previous commit.  I will get that done today.

taraspos commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -245,8 +241,7 @@ func handleGetReplicaCount(w http.ResponseWriter, r *http.Request) {
+	deployment, exists := getDeploymentFromCache(namespace, deploymentName)
 	if !exists {
 		writeJSONError(w, apiError{
 			Message: "Deployment not found in cache",
```
```
➜  k8s-deployment-scaler git:(development/helm) curl -X GET "https://localhost:8443/replica-count?namespace=default&deployment=lallala" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
{"message":"Deployment not found in cache","code":404}
```

This is the error I get when trying to query deployment that doesn't exist. 
Clients doesn't really need to know about the cache, just if deployment exist or not.

dereknorrbom replied:
Agreed, I will update this.

taraspos commented on helm/k8s-deployment-scaler/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+        - containerPort: 8443
+        readinessProbe:
+          tcpSocket:
+            port: 8443
```
Readiness probe is TCP based on HTTPS port, as results logs are full of:

```
2024/06/27 14:11:48 http: TLS handshake error from 10.244.0.1:36052: EOF
2024/06/27 14:11:53 http: TLS handshake error from 10.244.0.1:36054: EOF
2024/06/27 14:11:58 http: TLS handshake error from 10.244.0.1:60312: EOF
2024/06/27 14:12:03 http: TLS handshake error from 10.244.0.1:60326: EOF
2024/06/27 14:12:08 http: TLS handshake error from 10.244.0.1:51634: EOF
```

You should either handle EOF from health-checker, or come up with a different way to check health.

dereknorrbom replied:
I will suppress the EOF error to avoid adding additional complexity or adding any non-mTLS endpoints.

taraspos commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -245,8 +241,7 @@ func handleGetReplicaCount(w http.ResponseWriter, r *http.Request) {
 		return
```
Depending on the POST/GET request with missing query params error is different.
Validation logic is the same, it's a good candidate to move into separate function.


```
➜  k8s-deployment-scaler git:(development/helm) curl -X GET "https://localhost:8443/replica-count" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem                             
{"message":"Both namespace and deployment must be specified","code":400}
```

```
➜  k8s-deployment-scaler git:(development/helm) curl -X POST "https://localhost:8443/replica-count" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem                           
{"message":"Missing query parameters","code":400}
```

dereknorrbom replied:
I agree, I will create a `validateQueryParams` function.

taraspos commented on helm/k8s-deployment-scaler/templates/role.yaml:
```diff
@@ -0,0 +1,9 @@
+kind: Role
+metadata:
+  name: k8s-deployment-scaler-role
+  namespace: default
```
There is a bit of inconsistency between resource definitions, some of them specify namespace and other `default`.

In any case, `default` namespace doesn't seem like the best place to run kubernetes controller like this.

dereknorrbom replied:
role.yaml has been removed, it became YAGNI. I will move away from using default.

evanfreed reviewed (commented):
Following the readme trying to update coredns looks like it fails but succeeds for k8s-deployment-scaler.  Based on the design, I would suspect any deployment the service can list it can also manage or is this intended?

```
➜ curl -X GET "https://localhost:8443/replica-count?namespace=kube-system&deployment=coredns" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
{"replicaCount":3}

➜ curl -X POST -H "Content-Type: application/json" -d '{"replicas": 2}' "https://localhost:8443/replica-count?namespace=kube-system&deployment=coredns" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem

{"message":"Failed to update deployment","code":500}

➜ curl -X POST -H "Content-Type: application/json" -d '{"replicas": 2}' "https://localhost:8443/replica-count?namespace=default&deployment=k8s-deployment-scaler" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
{"replicaCount":2}

```

dereknorrbom committed:
clusterrole verbs update

Fix for https://github.com/dereknorrbom/k8s-deployment-scaler/pull/3#pullrequestreview-2145667379

- Add update to clusterrole.yaml verbs

dereknorrbom commented:
> Following the readme trying to update coredns looks like it fails but succeeds for k8s-deployment-scaler. Based on the design, I would suspect any deployment the service can list it can also manage or is this intended?
> 
> ```
> ➜ curl -X GET "https://localhost:8443/replica-count?namespace=kube-system&deployment=coredns" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
> {"replicaCount":3}
> 
> ➜ curl -X POST -H "Content-Type: application/json" -d '{"replicas": 2}' "https://localhost:8443/replica-count?namespace=kube-system&deployment=coredns" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
> 
> {"message":"Failed to update deployment","code":500}
> 
> ➜ curl -X POST -H "Content-Type: application/json" -d '{"replicas": 2}' "https://localhost:8443/replica-count?namespace=default&deployment=k8s-deployment-scaler" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
> {"replicaCount":2}
> ```

I just pushed an update that resolves this.  Please pull the change and redeploy with `make helm-upgrade`

```
➜  make helm-upgrade
Upgrading Helm chart...
helm upgrade k8s-deployment-scaler ./helm/k8s-deployment-scaler
Release "k8s-deployment-scaler" has been upgraded. Happy Helming!
NAME: k8s-deployment-scaler
LAST DEPLOYED: Thu Jun 27 09:07:21 2024
NAMESPACE: default
STATUS: deployed
REVISION: 2
TEST SUITE: None

➜  curl -X POST -H "Content-Type: application/json" -d '{"replicas": 2}' "https://localhost:8443/replica-count?namespace=kube-system&deployment=coredns" --cert ./certs/client-cert.pem --key ./certs/client-key.pem --cacert ./certs/ca-cert.pem
{"replicaCount":2}
```

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -25,26 +28,22 @@ import (
-		config, err = clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
-	}
-
+	config, err = getKubernetesConfig()
```
```suggestion
	config, err := getKubernetesConfig()
```

and remove the vars above. Looks leftover from earlier changes.

dereknorrbom replied:
Corrected. Thank you.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -310,8 +304,10 @@ func handlePostReplicaCount(w http.ResponseWriter, r *http.Request) {
-	_, err := clientset.AppsV1().Deployments(namespace).Update(context.TODO(), info.Deployment, metav1.UpdateOptions{})
+	ctx, cancel := context.WithTimeout(r.Context(), 10*time.Second)
+	defer cancel()
+	deployment.Spec.Replicas = &reqBody.Replicas
```
According to the docs. Deployment objects returned from the cache should be treated as read-only https://pkg.go.dev/k8s.io/client-go@v0.30.2/listers/apps/v1#DeploymentNamespaceLister.

dereknorrbom replied:
Thank you for the clarification, deployments are now retrieved from cache as well.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -310,8 +304,10 @@ func handlePostReplicaCount(w http.ResponseWriter, r *http.Request) {
+	ctx, cancel := context.WithTimeout(r.Context(), 10*time.Second)
+	defer cancel()
+	deployment.Spec.Replicas = &reqBody.Replicas
+	_, err := clientset.AppsV1().Deployments(namespace).Update(ctx, deployment, metav1.UpdateOptions{})
```
Feel free to use the [UpdateScale](https://pkg.go.dev/k8s.io/client-go@v0.30.2/listers/apps/v1#DeploymentNamespaceLister) method if you don't want to clone the deployment object.

dereknorrbom replied:
Great idea. Implemented.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -335,13 +331,26 @@ func listDeployments(w http.ResponseWriter, r *http.Request) {
+		list, err = deploymentLister.List(labels.Everything())
+	} else {
+		list, err = deploymentLister.Deployments(namespace).List(labels.Everything())
+	}
```
```suggestion
	list, err := deploymentLister.Deployments(namespace).List(labels.Everything())
```

The lister will returns all deployments if namespace is `""`.

dereknorrbom replied:
Thank you, I have implemented your suggestion.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -25,26 +28,22 @@ import (
-}
+var (
+	deploymentLister  appslisters.DeploymentLister
+	deploymentsSynced cache.InformerSynced
```
This variable doesn't appear to be needed globally.

dereknorrbom replied:
You are correct, updated. Thank you.

logand22 commented on helm/k8s-deployment-scaler/templates/clusterrole.yaml:
If you choose to use deployment scale instead of update you might want to update this.

dereknorrbom replied:
Thank you, updated.

logand22 commented on helm/k8s-deployment-scaler/templates/role.yaml:
Do you need a clusterrole and a role?

dereknorrbom replied:
No, I will remove the role and keep the clusterrole

dereknorrbom committed:
PR Comment Updates

- Update unit tests
- Remove unnecessary role and rolebinding
- `deploymentLister` and `deploymentsSynced` don't need to be global
- Directly use `deploymentLister.Deployments(namespace).List(labels.Everything())`
- Create a new `autoscalingv1.Scale` object with the desired replica count
- Use `clientset.AppsV1().Deployments(namespace).UpdateScale()` to update only the scale of the deployment
- Update clusterrole for deployment scale
- Remove the redundant variable declarations for config and err at the beginning of the main function
- Ensure all read requests come from cache

dereknorrbom committed:
Updates for PR comments

- Reorganized the project structure for improved maintainability and better logical organization
- Update cache-related errors
- Create validateQueryParams function for consistent error handling
- Relocate application from default namespace to a dedicated custom namespace
- Update README
- Relocate tests to be adjacent to their respective packages in @internal
- Implemented custom logging to suppress EOF errors
— Addresses readiness probe issues while maintaining mTLS-only endpoints
— Avoids complexity of handling certificates for health checks

dereknorrbom commented:
All updates per the comments have been pushed.
